portfolio/src/main/java/com/google/sps/servlets/ImageFormHandlerServlet.java

```diff
@@ -0,0 +1,98 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+/**
+ * When the user submits the form, Blobstore processes the file upload and then forwards the request
+ * to this servlet. This servlet can then process the request using the file URL we get from
+ * Blobstore, but in this case the image URL is put into an array to be used to appear on the webpage.
+ */
+@WebServlet("/my-form-handler")
+public class ImageFormHandlerServlet extends HttpServlet {
+  private List<String> images = new ArrayList<>();
+  //Variable to be used to allow multiple images to be uploaded,
+  // and allow the image on the page to refresh 
+  private int maxNumberOfImages = 2; 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(images));
+  }
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String imageUrl = getUploadedFileUrl(request, "image");
+    images.add(imageUrl);
+
+    if(images.size() >= maxNumberOfImages){
+      images.remove(0);
+    }
+    
+    response.sendRedirect("/index.html");
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
```

Comment:
Why check for null only to return it? Does isEmpty() perform a null check?

Reply:
isEmpty just checks if the array is empty and this is some of the sample code they provided which I used. Returning null, tells the function that not an appropriate image file was uploaded.

Reply:
Ok. Still, given this is simply returning null if blobkeys is null, this check, as is, isn't adding much value.

Looking deeper into this, blobkeys is never instantiated to an empty list. Because of this, blobkeys.isEmpty() can't be true. I encourage you to determine what happens if blobs.get() fails and reduce this check to focus on that.

Reply:
blobkeys can be in instantiated to an empty list, if you press the submit button without uploading a file